dotnet/src/SemanticKernel.Abstractions/Memory/AIContextBehavior.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.AI;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Base class for all AI context behaviors.
+/// </summary>
+/// <remarks>
+/// An AI context behavior is a component that can be used to enhance the AI's context management.
+/// It can listen to changes in the conversation, provide additional context to
+/// the AI model just before invocation and supply additional tools for function invocation.
+/// </remarks>
+[Experimental("SKEXP0130")]
+public abstract class AIContextBehavior
+{
+    /// <summary>
+    /// Called just after a new thread is created.
+    /// </summary>
+    /// <remarks>
+    /// Implementers can use this method to do any operations required at the creation of a new thread.
+    /// For example, checking long term storage for any data that is relevant to the current session based on the input text.
+    /// </remarks>
+    /// <param name="threadId">The ID of the new thread, if the thread has an ID.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>A task that completes when the context has been loaded.</returns>
+    public virtual Task OnThreadCreatedAsync(string? threadId, CancellationToken cancellationToken = default)
```

Comment:
I see some different names for the functions below, might suggest use a consistent naming.

Suggest instead of having `New`, use `Create` 
Suggest instead of having `Created`, use `Create` 

`OnNewMessage` -> `OnMessageCreate`

```suggestion
    public virtual Task OnThreadCreateAsync(string? threadId, CancellationToken cancellationToken = default)
```

Reply:
The present-tense past-tense naming is to indicate to users when the event is firing.  E.g. Created means that this event is firing after the thread was Created.  OnThreadDeleteAsync on the other hand is firing before the thread is deleted. This is a fairly common pattern elsewhere in .net too:
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1713

We may want to use OnThreadDeletingAsync though

Not sure about OnMessageCreated.  It's just a message passed to us, assuming anything else about it is probably a step too far.

Reply:
Updated to use rules using present continuous for before and past tense for after as described in rules and agreed in meeting.  Also updated OnNewMessage to MessageAdding to match this pattern.